FEXCore/Source/Interface/IR/Passes/x87StackOptimizationPass.cpp

```diff
@@ -0,0 +1,1148 @@
+  if (Offset != 0) {
+    OffsetTop = IREmit->_And(OpSize::i32Bit, IREmit->_Add(OpSize::i32Bit, OffsetTop, GetConstant(Offset)), GetConstant(7));
+    // GetTopWithCache_Slow already sets the cache so we don't need to set it here for offset == 0
+    TopOffsetCache[Offset] = OffsetTop;
+  }
+
+  return OffsetTop;
+}
+
+
+inline void X87StackOptimization::SetTopWithCache_Slow(Ref Value) {
+  IREmit->_StoreContext(1, GPRClass, Value, offsetof(FEXCore::Core::CPUState, flags) + FEXCore::X86State::X87FLAG_TOP_LOC);
+  InvalidateTopOffsetCache();
+  TopOffsetCache[0] = Value;
+}
+
+inline void X87StackOptimization::SetX87ValidTag(Ref Value, bool Valid) {
+  Ref AbridgedFTW = IREmit->_LoadContext(1, GPRClass, offsetof(FEXCore::Core::CPUState, AbridgedFTW));
+  Ref RegMask = IREmit->_Lshl(OpSize::i32Bit, GetConstant(1), Value);
+  Ref NewAbridgedFTW = Valid ? IREmit->_Or(OpSize::i32Bit, AbridgedFTW, RegMask) : IREmit->_Andn(OpSize::i32Bit, AbridgedFTW, RegMask);
+  IREmit->_StoreContext(1, GPRClass, NewAbridgedFTW, offsetof(FEXCore::Core::CPUState, AbridgedFTW));
+}
+
+inline Ref X87StackOptimization::GetX87ValidTag_Slow(uint8_t Offset) {
+  Ref AbridgedFTW = IREmit->_LoadContext(1, GPRClass, offsetof(FEXCore::Core::CPUState, AbridgedFTW));
+  return IREmit->_And(OpSize::i32Bit, IREmit->_Lshr(OpSize::i32Bit, AbridgedFTW, GetOffsetTopWithCache_Slow(Offset)), GetConstant(1));
+}
+
+inline Ref X87StackOptimization::LoadStackValueAtTop_Slow() {
+  return LoadStackValueAtOffset_Slow(0);
+}
+
+inline Ref X87StackOptimization::LoadStackValueAtOffset_Slow(uint8_t Offset) {
+  return IREmit->_LoadContextIndexed(GetOffsetTopWithCache_Slow(Offset), ReducedPrecisionMode ? 8 : 16, MMBaseOffset(), 16, FPRClass);
+}
+
+inline void X87StackOptimization::StoreStackValueAtTop_Slow(Ref Value, bool SetValid) {
+  StoreStackValueAtOffset_Slow(Value, 0, SetValid);
+}
+
+inline void X87StackOptimization::StoreStackValueAtOffset_Slow(Ref Value, uint8_t Offset, bool SetValid) {
+  OrderedNode* TopOffset = GetOffsetTopWithCache_Slow(Offset);
+  // store
+  IREmit->_StoreContextIndexed(Value, TopOffset, ReducedPrecisionMode ? 8 : 16, MMBaseOffset(), 16, FPRClass);
+  // mark it valid
+  // In some cases we might already know it has been previously set as valid so we don't need to do it again
+  if (SetValid) {
+    SetX87ValidTag(TopOffset, true);
+  }
+}
+
+inline Ref X87StackOptimization::RotateRight8(uint32_t V, Ref Amount) {
+  return IREmit->_Lshr(OpSize::i32Bit, GetConstant(V | (V << 8)), Amount);
+}
+
+inline std::optional<X87StackOptimization::StackMemberInfo> X87StackOptimization::MigrateToSlowPath_IfInvalid(uint8_t Offset) {
+  const auto& [Valid, StackMember] = StackData.top(Offset);
+  MigrateToSlowPathIf(Valid != StackSlot::VALID);
+  if (Valid == StackSlot::VALID) {
+    return StackMember;
+  }
+  return {};
+}
+
+inline Ref X87StackOptimization::LoadStackValue(std::optional<StackMemberInfo>& StackValue, uint8_t Offset) {
```

Comment:
Can we change the signature of this to simply

```
inline Ref X87StackOptimization::LoadStackValue(uint8_t Offset) {
```

and then include a call of `MigrateToSlowPath_IfInvalid` inside? Then all the Migrate calls goa way

Reply:
Sounds good. Only one case I left as is because we needed the value from `Migrate...`.